compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEInsertDmaOutOfOrderBlock.cpp

```diff
@@ -0,0 +1,202 @@
+  int8_t newRepeatCount = 0;
+  SmallVector<Value> newConnections;
+  SmallVector<SmallVector<Operation *>> lockAndBdOps;
+  for (AMDAIE::DMAStartOp s2mmDmaStartOp : s2mmDmaStartOps) {
+    // Reject the merge if the DMA start op already has more than one connection
+    // or if it is not packet-switched.
+    SmallVector<Value> connections = s2mmDmaStartOp.getConnections();
+    newConnections.append(connections.begin(), connections.end());
+    if (connections.size() != 1) {
+      return s2mmDmaStartOp.emitOpError()
+             << "expected exactly one connection for DMA start op, but found "
+             << connections.size() << "\n";
+    }
+    AMDAIE::ConnectionOp connectionOp =
+        dyn_cast<AMDAIE::ConnectionOp>(connections[0].getDefiningOp());
+    if (connectionOp.getConnectionType() != AMDAIE::ConnectionType::Packet) {
+      return s2mmDmaStartOp.emitOpError()
+             << "out-of-order mode requires DMA to be packet-switched\n";
+    }
+    // Find the corresponding MM2S DMA start op.
+    // TODO(zhewen): Add support for npu.write_bd when the MM2S channel is
+    // located in the shim tile.
+    FailureOr<AMDAIE::DMAStartOp> maybeMm2sDmaStartOp =
+        getMm2sDmaStart(connectionOp, mm2sChannelToDmaStarts);
+    if (failed(maybeMm2sDmaStartOp)) {
+      return s2mmDmaStartOp.emitOpError()
+             << "could not find corresponding MM2S DMA start op";
+    }
+    // Get the blocks in the S2MM and MM2S DMA start ops.
+    SmallVector<Block *> s2mmBlocks = llvm::to_vector(llvm::map_range(
+        s2mmDmaStartOp.getBody().getBlocks(), [](Block &b) { return &b; }));
+    SmallVector<Block *> mm2sBlocks = llvm::to_vector(
+        llvm::map_range(maybeMm2sDmaStartOp.value().getBody().getBlocks(),
+                        [](Block &b) { return &b; }));
+    int8_t numBDs = 0;
+    for (size_t i = 0; i < s2mmBlocks.size(); i++) {
+      lockAndBdOps.push_back({});
+      std::optional<uint32_t> maybeS2mmBdId;
+      for (Operation &op : *s2mmBlocks[i]) {
+        // Find the DMA BDs and locks in the S2MM block.
+        if (isa<AMDAIE::DMABDOp, AMDAIE::UseLockOp>(op))
+          lockAndBdOps.back().push_back(&op);
+        if (auto bdOp = dyn_cast<AMDAIE::DMABDOp>(op)) {
+          numBDs++;
+          maybeS2mmBdId = bdOp.getBdId();
+        }
+      }
+      if (lockAndBdOps.back().size() == 0) {
+        lockAndBdOps.pop_back();
+        continue;
+      }
+      if (!maybeS2mmBdId.has_value()) {
+        return s2mmDmaStartOp.emitOpError()
+               << "BD ID is not assigned in S2MM block";
+      }
+      // Embed the S2MM BD ID in the corresponding MM2S packer header.
+      auto maybeMm2sDmaBdPacketOps =
+          mm2sBlocks[i]->getOps<AMDAIE::DmaBdPacketOp>();
+      if (llvm::range_size(maybeMm2sDmaBdPacketOps) != 1) {
+        return maybeMm2sDmaStartOp.value().emitOpError()
+               << "expected exactly one `amdaie.dma_bd_packet` op in each MM2S "
+                  "block";
+      }
```

Comment:
Maybe this could be put in the DMAStartOp verifier? And maybe the DMAStartOp could then have a utility to retrieve the `amdaie.dma_bd_packet` ops? This could potentially simplify the logic here and above. WDYT?